src/kernel/include/thd.h

```diff
@@ -441,6 +465,42 @@ thd_preemption_state_update(struct thread *curr, struct thread *next, struct pt_
 	memcpy(&curr->regs, regs, sizeof(struct pt_regs));
 }
 
+static inline void 
+thd_rcvcap_pending_deliver(struct thread *thd, struct pt_regs *regs) {
+	unsigned long a = 0, b = 0;
+	int all_pending = thd_rcvcap_all_pending_get(thd);
+
+	thd_state_evt_deliver(thd, &a, &b);
+	if (all_pending) {
+		__userregs_setretvals(regs, thd_rcvcap_all_pending(thd), a, b);
+	} else {
+		thd_rcvcap_pending_dec(thd);
+		__userregs_setretvals(regs, thd_rcvcap_pending(thd), a, b);
+	}
+}
+
+static int
+thd_switch_update(struct thread *thd, struct pt_regs *regs, int issame)
```

Comment:
I don't know 1. where all this is called from, nor 2. how many times it is called.  If it is on a fast-path, and not called that often, `inline` might be appropriate.  Up to you.